lib/OpenQA/WebAPI/Controller/Step.pm

```diff
@@ -274,9 +310,14 @@ sub _new_screenshot ($self, $tags, $image_name, $matches = undef) {
     return \%screenshot;
 }
 
-sub _basic_needle_info ($self, $name, $distri, $version, $file_name, $needles_dir) {
+sub _basic_needle_info (
+    $self, $name, $distri, $version, $file_name, $needles_dir,
+    $needle_ref = undef,
+    $needle_url = undef
+  )
+{
     $file_name //= "$name.json";
-    $file_name = locate_needle($file_name, $needles_dir) if !-f $file_name;
+    $file_name = locate_needle($file_name, $needles_dir, $needle_ref, $needle_url) if !-f $file_name;
```

Comment:
Now line 332 needs to be changed:
```diff
-     $needle->{json} = $file_name;
+     $needle->{json} = "$file_name";
```
because that is a Mojo::File object and Cpanel::JSON::XS doesn't like serializing objects...

Reply:
Alternatively we can use
```
Cpanel::JSON::XS->new->convert_blessed->pretty
```
in `templates/webapi/step/edit.html.ep`